strutil/limbuffer.go

```diff
@@ -0,0 +1,58 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package strutil
+
+import (
+	"bytes"
+)
+
+type LimitedBuffer struct {
+	buffer   *bytes.Buffer
+	maxLines int
+	maxBytes int
+}
+
+func NewLimitedBuffer(maxLines, maxBytes int) *LimitedBuffer {
+	return &LimitedBuffer{
+		buffer:   &bytes.Buffer{},
+		maxLines: maxLines,
+		maxBytes: maxBytes,
+	}
+}
+
+func (wr *LimitedBuffer) Write(data []byte) (int, error) {
+	sz, err := wr.buffer.Write(data)
```

Comment:
It makes me uneasy that we write into buffer here without checking the size of data. I guess for our use-case this is not a problem and probably not a problem in general. But my pedantic side does not like it that this might grow wr.buffer well beyond anything it was limited to. I understand the reasoning (much simpler this way).

Reply:
Thank you for this comment! It made me realize there is better way of doing this logic (and not really complicated, in fact I think Maciej suggested that too in one of his comments), so I've rewritten it to never allow such case by splitting the logic into three cases, truncating the buffer upfront to make space and just append to it. There is probably a couple of ways to implement this logic... Let me know how you like it in the new shape.